### ISSUE TITLE
SameAs should warn if used with value types

### ISSUE THREAD

**Issue by jnm2** [nunit/nunit.analyzers#276]
Motivation: https://github.com/nunit/nunit/issues/3614#issuecomment-669962745

If either the expected value or actual value or both are a value-typed expression, `Assert.AreSame` or `Is.SameAs` are guaranteed to fail because C# will implicitly convert the value type to a new boxed instance. It doesn't make sense to use AreSame/SameAs unless the types of the passed expressions are both reference types.

It would make sense to offer a fix to change SameAs to EqualTo since that was probably the intent.

Should show an error or warning on `SameAs`:

```cs
class C
{
    void M()
    {
        var x = 3;
        Assert.That(x, Is.SameAs(3));

        // Also this
        Assert.That((object)x, Is.SameAs(3));

        // Also this
        Assert.That(x, Is.SameAs((object)3));
    }
}
```

Should not show an error or warning because the expressions are both reference-typed:

```cs
class C
{
    void M()
    {
        var x = 3;
        Assert.That((object)x, Is.SameAs((object)3));
    }
}
```

**Comment by manfred-brands** [comment]
I will work on this over the weekend.

**Comment by manfred-brands** [comment]
Analyzer and UnitTest done. Code Fix pending.

**Comment by manfred-brands** [comment]
Code Fix and documentation done.
